Nouvelle API pour gérer les contenus

artragis opened the pull request:
le but de cette PR est de créer une API pour gérer les tutos (pour l'instant, on ne joue pas pour l'affichage des publications).

Cela peut être vu comme le pendant backend de la nouvelle interface de rédaction qui est en cours de maquettage.

Je vais faire deux endpoints ou groupe de endpoints :

- [x] Gestion des chapitre/extraits/parties. 
- [x] Gestion des métadata (titre, licence, catégorie...)

Ça serait bien que j'ai un petit descriptif de ce que je dois mettre dans ces endpoints pour correspondre aux besoins du frontend

# QA

- Review
- si vous voyez des cas qui ne sont pas testés dans `test_api.py` dites-le je les ajoute.

artragis committed:
Nouvelle API pour gérer les contenus

Cette API permet d'ajouter, trier, modifier des extraits et des espaces hiérarchiques.
Elle est un préambule au travail de génération d'un nouvel éditeur de tutoriel sur le site.

motet-a commented on zds/tutorialv2/urls/urls_contents.py:
```diff
@@ -28,14 +28,24 @@
+        r'(?P<parent_container_slug>[a-zA-Z0-9_-]+)/(?P<container_slug>[a-zA-Z0-9_-]+)/$',
+        DisplayContainer.as_view(public_is_prioritary=False, template_name='tutorialv2/view/container_nosplit.html'),
+        name='view-container-nosplit'),
+    url(r'^nosplit/(?P<pk>\d+)/(?P<slug>[a-zA-Z0-9_-]+)/$',
```
Faudrait peut-être renommer ces URLs. Tous les chemins sont en français actuellement et ce “nosplit” fait un peu tâche je trouve.

artragis replied:
ça va surtout être viré :)

artragis replied:
ça a été viré

motet-a commented on zds/tutorialv2/api/serializers.py:
```diff
@@ -0,0 +1,163 @@
+        :return:
+        """
+
+        error = not super(ChildrenListSerializer, self).is_valid(raise_exception)
```
Ça ne serait pas `super().is_valid(raise_exception)` par hazard ?

artragis replied:
en python 3 oui, mais j'ai commencé ce code en py2 :)

motet-a commented on zds/tutorialv2/api/serializers.py:
```diff
@@ -0,0 +1,163 @@
+        if not self._validated_data['extracts'] and not self._validated_data['containers']:
+            error = True
+            messages['extracts'] = _('Le contenu semble vide.')
+        if raise_exception and error:
```
J’ai un peu du mal à voir l’utilité de la variable `error`. On ne peut pas faire `if raise_exception and messages != {}:` ?

artragis replied:
on peut, j'essayais de me dépatouiller avec le fait que je n'utilise pas un sérializer standard car ce n'est pas un modèle en bdd.

motet-a commented on zds/tutorialv2/mixins.py:
```diff
@@ -129,7 +129,7 @@ def get_versioned_object(self):
             if versioned.slug != slug:
                 if slug != self.object.slug:  # retro-compatibility, but should raise permanent redirect instead
-                    raise Http404("Ce slug n'existe pas pour ce contenu.")
+                    raise Http404("Ce slug n'existe pas pour ce contenu.{} vs {}".format(slug, versioned.slug))
```
Je crois qu’il y a un `.` en trop. Ça serait bien d’avoir un message un peu plus détaillé, je pense que je ne comprendrais pas trop ce qu’il se passe si je voyais ça :upside_down_face:

artragis replied:
Non le . est volontaire. Et ok je peux mettre "expected = {}, got {}". Notons que c'estu n message de debug.

motet-a replied:
Aaaaah OK. Mais comme il manque un espace après le `.`, je me suis dit que c’était peut-être une erreur :upside_down_face:

artragis replied:
la typographie n'est pas mon dada.

Commit:
start work for metadata API endpoints

Commit:
Ajoute un endpoint d'API pour les catégories de tuto.

En effet en corrigeant les problèmes des sérializers je me suis rendu compte
que cela est nécessaire car ces derniers demandent à envoyer les pk des catégories
pour créer/supprimer le contenu.

Notons que le même problème se posera pour les licences mais semble pouvoir être corrigé rapidement.

Anto59290 commented on zds/tutorialv2/api/permissions.py:
```diff
@@ -0,0 +1,45 @@
+            request_param_user = request.parser_context['kwargs'].get('user', '0')
+            current_user = request.user
+            return current_user and current_user.pk == int(request_param_user)
+        except ValueError:  # not an int
```
On doit pouvoir sortir les deux première lignes du try/except il me semble.

artragis replied:
je ne vois pas trop l'intérêt.

motet-a replied:
L’intérêt c’est d’éviter d’attraper des exceptions par erreur. Par exemple, si `request.parser_context['kwargs'].get('user', '0')` déclenche un étrange accesseur de propriété buggé qui lance une `ValueError`, ça sera attrapé par le `except` comme si de rien n’était.

Il vaut mieux mettre le strict minimum dans le bloc du `try:`.

artragis replied:
ok je vais le faire, mais bon, ça m'embête de casser la lecture pour un truc improbable.

Anto59290 commented on zds/tutorialv2/api/serializers.py:
```diff
@@ -0,0 +1,255 @@
+from rest_framework.fields import CharField, empty
+
+from zds.tutorialv2.api.view_models import ChildrenViewModel, ChildrenListViewModel, UpdateChildrenListViewModel
+from gettext import gettext as _
```
Est ce qu'il ne vaut mieux pas importer depuis from `django.utils.translation` pour rester cohérent avec le reste du code ?

artragis replied:
ah si,

artragis commented:
J'enlève le wip car je ne vois pas quoi ajouter pour l'instant.

Anto59290 commented on zds/tutorialv2/api/serializers.py:
```diff
@@ -0,0 +1,262 @@
+            try:
+                return func(*args, **kwargs)
+            except exception1:
+                logger.warning('Error translated fril %s to %s', exception1, exception2(message))
```
`fril` --> `from` ;)

Anto59290 commented on zds/tutorialv2/api/serializers.py:
```diff
@@ -0,0 +1,262 @@
+    """
+
+    def update(self, instance, validated_data):
+        pass
```
Pourquoi a-t-on besoin de cette méthode ? Si on souhaite la garder est ce qu'une `NotImplementedError` ne serait pas mieux qu'un pass silencieux ?

artragis replied:
il est probable que ça soit l'autogénerate de pycharm.

Anto59290 commented on zds/tutorialv2/api/view_models.py:
```diff
@@ -0,0 +1,79 @@
+
+    def __init__(self, extracts=None, containers=None, introduction=None, conclusion=None, authors=None,
+                 remove_deleted_children=False, message='', **__):
+        super(UpdateChildrenListViewModel, self).__init__(extracts, containers, introduction, conclusion, authors)
```
J'imagine que tu as du commencer en Python 2.X, mais du coup plus besoin de l'argument dans le super

artragis replied:
j'ai effectivement commencé en 2.7, je modifie ça.

Anto59290 reviewed (commented):
J'ai fait une première passe de relecture, j'ai eu pas mal de difficultés a comprendre la méthode is_valid de ChildrenListSerializer.

artragis commented:
>  j'ai eu pas mal de difficultés a comprendre la méthode is_valid de ChildrenListSerializer.

Le principal problème de ce serializer, c'est qu'il n'est pas basé sur un modèle stocké en bdd (puisque c'est du git pur). Du coup il faut tout réécrire à la main.

A la base j'avais voulu rendre la génération d'un contenu générique en s'occupant aussi bien de la création des parties/chapitres que des textes du tuto en soi.

Du coup tu as une disjonctions à faire : soit on est face à une liste de `extracts` soit face à une liste de `containers`.

Ensuite, il faut s'assurer qu'il n'y a pas de doublon au sein des demandes d'ajout.

Enfin, il faut s'assurer que chaque container a la bonne forme.

De plus, puisque le serializer gère aussi bien l'ajout d'une partie/extrait que le réordonnancement complet desdits extraits, il faut pouvoir utiliser un flag pour dire "ne fait qu'ajouter à la suite" ou bien "force l'ordre que je te donne".

En soi, je pense qu'il faudrait que je refactore un peu cette fonction pour faire apparaître les étapes de manière plus claire.

artragis committed:
py3 style

artragis commented:
Quelqu'un pour faire la QA svp?

Situphen commented:
Je l'ai déployée sur la bêta mais j'ai 2 erreurs 500 sur `/api`.

```
[ERROR] -- 2019-08-20 07:03:13,216 -- django.request : Internal Server Error: /api/
Traceback (most recent call last):
  File "/opt/zds/virtualenv/lib/python3.5/site-packages/django/core/handlers/exception.py", line 34, in inner
    response = get_response(request)
  File "/opt/zds/virtualenv/lib/python3.5/site-packages/django/core/handlers/base.py", line 126, in _get_response
    response = self.process_exception_by_middleware(e, request)
  File "/opt/zds/virtualenv/lib/python3.5/site-packages/django/core/handlers/base.py", line 124, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/opt/zds/virtualenv/lib/python3.5/site-packages/django/views/decorators/csrf.py", line 54, in wrapped_view
    return view_func(*args, **kwargs)
  File "/opt/zds/virtualenv/lib/python3.5/site-packages/django/views/generic/base.py", line 68, in view
    return self.dispatch(request, *args, **kwargs)
  File "/opt/zds/virtualenv/lib/python3.5/site-packages/rest_framework/views.py", line 495, in dispatch
    response = self.handle_exception(exc)
  File "/opt/zds/virtualenv/lib/python3.5/site-packages/rest_framework/views.py", line 455, in handle_exception
    self.raise_uncaught_exception(exc)
  File "/opt/zds/virtualenv/lib/python3.5/site-packages/rest_framework/views.py", line 492, in dispatch
    response = handler(request, *args, **kwargs)
  File "/opt/zds/virtualenv/lib/python3.5/site-packages/rest_framework_swagger/views.py", line 32, in get
    schema = generator.get_schema(request=request)
  File "/opt/zds/virtualenv/lib/python3.5/site-packages/rest_framework/schemas/generators.py", line 281, in get_schema
    links = self.get_links(None if public else request)
  File "/opt/zds/virtualenv/lib/python3.5/site-packages/rest_framework/schemas/generators.py", line 317, in get_links
    if not self.has_view_permissions(path, method, view):
  File "/opt/zds/virtualenv/lib/python3.5/site-packages/rest_framework/schemas/generators.py", line 390, in has_view_permissions
    view.check_permissions(view.request)
  File "/opt/zds/virtualenv/lib/python3.5/site-packages/rest_framework/views.py", line 334, in check_permissions
    if not permission.has_permission(request, self):
  File "/opt/zds/app/zds/member/api/permissions.py", line 46, in has_permission
    view.get_object())
  File "/opt/zds/virtualenv/lib/python3.5/site-packages/rest_framework/generics.py", line 85, in get_object
    queryset = self.filter_queryset(self.get_queryset())
  File "/opt/zds/virtualenv/lib/python3.5/site-packages/rest_framework/generics.py", line 68, in get_queryset
    % self.__class__.__name__
AssertionError: 'ExportView' should either include a `queryset` attribute, or override the `get_queryset()` method.
```

```
[ERROR] -- 2019-08-20 13:51:54,100 -- django.request : Internal Server Error: /api/
Traceback (most recent call last):
  File "/opt/zds/virtualenv/lib/python3.5/site-packages/django/core/handlers/exception.py", line 34, in inner
    response = get_response(request)
  File "/opt/zds/virtualenv/lib/python3.5/site-packages/django/core/handlers/base.py", line 126, in _get_response
    response = self.process_exception_by_middleware(e, request)
  File "/opt/zds/virtualenv/lib/python3.5/site-packages/django/core/handlers/base.py", line 124, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/opt/zds/virtualenv/lib/python3.5/site-packages/django/views/decorators/csrf.py", line 54, in wrapped_view
    return view_func(*args, **kwargs)
  File "/opt/zds/virtualenv/lib/python3.5/site-packages/django/views/generic/base.py", line 68, in view
    return self.dispatch(request, *args, **kwargs)
  File "/opt/zds/virtualenv/lib/python3.5/site-packages/rest_framework/views.py", line 495, in dispatch
    response = self.handle_exception(exc)
  File "/opt/zds/virtualenv/lib/python3.5/site-packages/rest_framework/views.py", line 455, in handle_exception
    self.raise_uncaught_exception(exc)
  File "/opt/zds/virtualenv/lib/python3.5/site-packages/rest_framework/views.py", line 492, in dispatch
    response = handler(request, *args, **kwargs)
  File "/opt/zds/virtualenv/lib/python3.5/site-packages/rest_framework_swagger/views.py", line 32, in get
    schema = generator.get_schema(request=request)
  File "/opt/zds/virtualenv/lib/python3.5/site-packages/rest_framework/schemas/generators.py", line 281, in get_schema
    links = self.get_links(None if public else request)
  File "/opt/zds/virtualenv/lib/python3.5/site-packages/rest_framework/schemas/generators.py", line 319, in get_links
    link = view.schema.get_link(path, method, base_url=self.url)
  File "/opt/zds/virtualenv/lib/python3.5/site-packages/rest_framework/schemas/inspectors.py", line 217, in get_link
    fields += self.get_serializer_fields(path, method)
  File "/opt/zds/virtualenv/lib/python3.5/site-packages/rest_framework/schemas/inspectors.py", line 360, in get_serializer_fields
    for field in serializer.fields.values():
  File "/opt/zds/virtualenv/lib/python3.5/site-packages/rest_framework/serializers.py", line 364, in fields
    self._fields[key] = value
  File "/opt/zds/virtualenv/lib/python3.5/site-packages/rest_framework/utils/serializer_helpers.py", line 147, in __setitem__
    field.bind(field_name=key, parent=self.serializer)
  File "/opt/zds/virtualenv/lib/python3.5/site-packages/rest_framework/fields.py", line 378, in bind
    (field_name, self.__class__.__name__, parent.__class__.__name__)
AssertionError: It is redundant to specify `source='extracts'` on field 'ListField' in serializer 'ChildrenListModifySerializer', because it is the same as the field name. Remove the `source` keyword argument.
```

Situphen committed:
Corrige quelques bouts de code

Situphen committed:
PEP 8 + Importations dans l'ordre alphabétique
